apiserver/common/leadership.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package common
+
+import (
+	"github.com/juju/errors"
+	"gopkg.in/juju/names.v2"
+
+	"github.com/juju/juju/apiserver/facade"
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/core/leadership"
+	"github.com/juju/juju/state"
+)
+
+//go:generate mockgen -package mocks -destination mocks/leadership.go github.com/juju/juju/apiserver/common LeadershipPinningBackend,LeadershipMachine
+
+// LeadershipMachine is an indirection for state.machine.
+type LeadershipMachine interface {
+	ApplicationNames() ([]string, error)
+}
+
+type leadershipMachine struct {
+	*state.Machine
+}
+
+// LeadershipPinningBacked describes state method wrappers used by this API.
+type LeadershipPinningBackend interface {
+	Machine(string) (LeadershipMachine, error)
+}
+
+type leadershipPinningBackend struct {
+	*state.State
+}
+
+// Machine wraps state.Machine to return an implementation
+// of the LeadershipMachine indirection.
+func (s leadershipPinningBackend) Machine(name string) (LeadershipMachine, error) {
+	m, err := s.State.Machine(name)
+	if err != nil {
+		return nil, err
+	}
+	return leadershipMachine{m}, nil
+}
+
+// API exposes leadership pinning and unpinning functionality for remote use.
+type LeadershipPinningAPI interface {
+	PinMachineApplications() (params.PinApplicationsResults, error)
+	UnpinMachineApplications() (params.PinApplicationsResults, error)
+}
+
+// NewLeadershipPinningFacade creates and returns a new leadership API.
+// This signature is suitable for facade registration.
+func NewLeadershipPinningFacade(ctx facade.Context) (LeadershipPinningAPI, error) {
+	st := ctx.State()
+	model, err := st.Model()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	pinner, err := ctx.LeadershipPinner(model.UUID())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return NewLeadershipPinningAPI(leadershipPinningBackend{st}, model.ModelTag(), pinner, ctx.Auth())
+}
+
+// NewLeadershipPinningAPI creates and returns a new leadership API from the
+// input tag, Pinner implementation and facade Authorizer.
+func NewLeadershipPinningAPI(
+	st LeadershipPinningBackend, modelTag names.ModelTag, pinner leadership.Pinner, authorizer facade.Authorizer,
+) (LeadershipPinningAPI, error) {
+	return &leadershipPinningAPI{
+		st:         st,
+		modelTag:   modelTag,
+		pinner:     pinner,
+		authorizer: authorizer,
+	}, nil
+}
+
+type leadershipPinningAPI struct {
+	st         LeadershipPinningBackend
+	modelTag   names.ModelTag
+	pinner     leadership.Pinner
+	authorizer facade.Authorizer
+}
+
+// PinMachineApplications pins leadership for applications represented by units
+// running on the auth'd machine.
+func (a *leadershipPinningAPI) PinMachineApplications() (params.PinApplicationsResults, error) {
+	if !a.authorizer.AuthMachineAgent() {
+		return params.PinApplicationsResults{}, ErrPerm
+	}
+	return a.pinMachineAppsOps(a.pinner.PinLeadership)
```

Comment:
I did think about this. I can see how this might be generalised on the entity based on who is auth'd:
- Machine: applications for units on machine.
- Unit: unit's application.
- Application: application (obviously).

Either of the last 2 might come in handy for model generations...